### PR TITLE
HipTensor Benchmarking Overhaul

### PR DESCRIPTION
**Pull request by dlangbe**
Includes changes from [271](https://github.com/ROCm/hipTensor/pull/271) and [273](https://github.com/ROCm/hipTensor/pull/273) with additional fixes for printout formatting and controlled header printing.

### CODE REVIEW

File: library/src/hiptensor_options.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    void HiptensorOptions::setDefaultParams(bool val)
+    {
+        mUsingDefaultParams = val;
+    }
+
+    void HiptensorOptions::setValidation(std::string val)
+    {
+        if(val.compare("ON") == 0)
```
Comment (CongMa13) [inline library/src/hiptensor_options.cpp:99]: Could we convert `val` to capital before comparing with `std::toupper`? Then `val` could be `On`, `on` or `ON`

File: library/src/hiptensor_options.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    void HiptensorOptions::setOstream(std::string file)
+    {
+        mOstream.initializeStream(file);
+    }
+
+    void HiptensorOptions::setOmits(int mask)
+    {
+        if(mask & 1)
+            mOmitSkipped = true;
```
Comment (CongMa13) [inline library/src/hiptensor_options.cpp:56]: Should we set `mOmitSkipped` to false if `mask & 1 == 0`. 

Same for other options.

File: scripts/performance/BenchmarkPermutation.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# Copyright (C) 2022-2024 Advanced Micro Devices, Inc. All rights reserved.
+
+set -eux
+
+# ensure this script is in the cwd
+cd "$(dirname "${BASH_SOURCE[0]}")"
+
+output_dir=hiptensor-benchmarks
+build_dir=../../build/bin/
+config_dir=../../test/02_permutation/configs
```
Comment (CongMa13) [inline scripts/performance/BenchmarkPermutation.sh]: Just wondered if this should be `..../configs/bench`?

File: test/03_reduction/reduction_test.cpp [hunk 4]
```diff
@@ -359,94 +447,134 @@ namespace hiptensor
...
+                                           hipDataTypeSize(acDataType),
+                                           std::multiplies<size_t>());
+
+            size_t sizeCD = std::accumulate(extentC.begin(),
+                                            extentC.end(),
+                                            hipDataTypeSize(acDataType),
+                                            std::multiplies<size_t>());
+
+            mElapsedTimeMs        = float64_t(timeMs);
+            mTotalGFlops          = 2.0 * ((sizeA * sizeCD) / hipDataTypeSize(acDataType));
```
Comment (CongMa13) [inline test/03_reduction/reduction_test.cpp]: The flops of reduction is `sizeA / hipDataTypeSize(acDataType))`

Say there is a tensor A of shape [3,4,5], and the output B of reduction is [3,4]

The size of B is `3*4` and each entry is the summation of 5 entries of A. 

So, the summation is called `3*4*5` times.

File: library/src/hiptensor_options.cpp (hunk 2 above)
Comment (cgmillette) [reply library/src/hiptensor_options.cpp:56]: Agreed.

File: library/src/include/hiptensor_options.hpp [hunk 5]
```diff
@@ -49,16 +49,27 @@ namespace hiptensor
...
         ~HiptensorOptions()                  = default;
 
-        void parseOptions(int argc, char** argv);
+        void setOstream(std::string file);
         void setOmits(int mask);
+        void setDefaultParams(bool val);
+        void setValidation(std::string val);
+        void setHotRuns(int runs);
+        void setColdRuns(int runs);
+        void setInputFilename(std::string file);
```
Comment (cgmillette) [inline library/src/include/hiptensor_options.hpp]: We may want to be a bit more specific about what the files are. 
Suggest setInputYAMLFilename..
Suggest setOutputStreamFilename..

File: test/01_contraction/contraction_test.cpp [hunk 6]
```diff
@@ -670,102 +797,153 @@ namespace hiptensor
...
+            CHECK_HIP_ERROR(hipEventSynchronize(stopEvent))
 
-            // Compute tolerance based on compute type
-            auto dimension = a_ms_ks.mLengths.size() / 2;
-            auto nelems_k  = std::accumulate(a_ms_ks.mLengths.begin() + dimension,
-                                            a_ms_ks.mLengths.end(),
-                                            size_t{1},
-                                            std::multiplies<size_t>());
+            auto timeMs = 0.0f;
+            CHECK_HIP_ERROR(hipEventElapsedTime(&timeMs, startEvent, stopEvent));
```
Comment (cgmillette) [inline test/01_contraction/contraction_test.cpp:804]: Don't forget to destroy your events

**Review by cgmillette — COMMENTED** [review]
Changes are looking great so far. Should be good after the feedbacks are addressed

File: library/src/hiptensor_options.cpp (hunk 2 above)
Comment (dlangbe) [reply library/src/hiptensor_options.cpp:56]: All omit flags are initialized to false

Comment (cgmillette) [reply library/src/hiptensor_options.cpp:56]: Yes, however if you were ever to call this function again, the non-masked items need to be reset back to false.

**Review by cgmillette — DISMISSED** [review]
LGTM

**Review by cgmillette — APPROVED** [review]

**Pull request merged**